File: sections/21_graphs_i_mixed_contest/problems/e_station_coverage_report/solution.py
import sys
from collections import deque


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    n, m, k = data[:3]
    graph = [[] for _ in range(n)]
    index = 3
    for _ in range(m):
        u, v = data[index] - 1, data[index + 1] - 1
        index += 2
        graph[u].append(v)
        graph[v].append(u)
    distance = [-1] * n
    queue = deque()
    for station in data[index:index + k]:
        station -= 1
        if distance[station] == -1:
            distance[station] = 0
            queue.append(station)
    while queue:
        u = queue.popleft()
        for v in graph[u]:
            if distance[v] == -1:
                distance[v] = distance[u] + 1
                queue.append(v)
    unreachable = distance.count(-1)
    farthest = max(distance)
    vertex = distance.index(farthest) + 1
    print(unreachable, vertex, farthest)


if __name__ == "__main__":
    main()

File: sections/21_graphs_i_mixed_contest/problems/e_station_coverage_report/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m, k;
    cin >> n >> m >> k;
    vector<vector<int>> graph(n);
    while (m--) {
        int u, v;
        cin >> u >> v;
        --u; --v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    vector<int> distance(n, -1);
    queue<int> queue;
    while (k--) {
        int station;
        cin >> station;
        --station;
        if (distance[station] == -1) {
            distance[station] = 0;
            queue.push(station);
        }
    }
    while (!queue.empty()) {
        int u = queue.front();
        queue.pop();
        for (int v : graph[u]) if (distance[v] == -1) {
            distance[v] = distance[u] + 1;
            queue.push(v);
        }
    }
    int unreachable = 0, vertex = -1, farthest = -1;
    for (int i = 0; i < n; ++i) {
        if (distance[i] == -1) {
            ++unreachable;
        } else if (distance[i] > farthest) {
            farthest = distance[i];
            vertex = i;
        }
    }
    cout << unreachable << ' ' << vertex + 1 << ' ' << farthest << '\n';
}
